core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/sendsemantics/partitioners/ParallelBatchingPartitioner.scala

```diff
@@ -1,29 +1,70 @@
 package edu.uci.ics.amber.engine.architecture.sendsemantics.partitioners
 
+import edu.uci.ics.amber.engine.common.Constants
 import edu.uci.ics.amber.engine.common.ambermessage.{DataFrame, DataPayload, EndOfUpstream}
 import edu.uci.ics.amber.engine.common.tuple.ITuple
 import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
 
+import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
 abstract class ParallelBatchingPartitioner(batchSize: Int, receivers: Seq[ActorVirtualIdentity])
     extends Partitioner {
-  var batches: Array[Array[ITuple]] = _
-  var currentSizes: Array[Int] = _
 
+  // A bucket corresponds to a partition. When Reshape is not enabled, a bucket has just one receiver.
+  // Reshape divides a skewed partition onto multiple workers. So, with Reshape, a bucket can have
+  // multiple receivers. First receiver in the bucket is the original receiver for that partition.
+  val numBuckets = receivers.length
+  var bucketsToReceivers = new mutable.HashMap[Int, ArrayBuffer[ActorVirtualIdentity]]()
+  var receiverToBatch = new mutable.HashMap[ActorVirtualIdentity, Array[ITuple]]()
+  var receiverToCurrBatchSize = new mutable.HashMap[ActorVirtualIdentity, Int]()
```

Comment:
understood by reading the comments. however, I wonder if it is worth defining a `Bucket` class to organize the information. right now its information is saved in pieces in different hashmaps, which is fragile and not very efficient to write code.

Reply:
We can have a discussoin about this later. In general, the APIs of using maps and buckets is a common one. Actually I found that the flux paper is also doing something very similar.